algorithmic-strategies/problem-set-algorithmic-strategies.md

```diff
@@ -0,0 +1,336 @@
+# Problem Set: Algorithmic Strategies
+
+## Directions
+
+Complete all questions below.
+
+## Practice
+
+<!-- prettier-ignore-start -->
+### !challenge
+* type: code-snippet
+* language: python3.6
+* id: ff66fd15
+* title: Algorithmic Strategies
+### !question
+
+The **0-1 Knapsack Problem** poses a question of vital importance to cat burglars the world over! Assume we are in a room of \(n\) treasures. Each of the treasures has a weight and a value. We have a knapsack to put these items in, and we can carry up to `max_weight` units of weight. We're only a moderately successful cat burglar, so once we leave the room, an alarm will trigger, alerting security.
+
+We only have one chance to pull off this heist, and make our getaway!
+
+What is the maximum value of the items that we can fit in our knapsack?
+
+Notice that this problem does _not_ ask which items those are, although some variations will do so. **We are only interested in the total maximum value.**
+
+Sourced from: [GeeksforGeeks](https://www.geeksforgeeks.org/0-1-knapsack-problem-dp-10/)
+
+As an example, let's consider the case where there are three items:
+| Item | Weight | Value  |
+|------|--------|--------|
+| 1    | 10     | 60     |
+| 2    | 20     | 100    |
+| 3    | 30     | 120    |
+
+If we are able to carry 50 units of weight, what is the maximum value of the items we can fit in our knapsack?
+
+In this case, we can take items 1 and 2, which have a combined weight of 50, and a combined value of 220. So our maximum value is 220.
+
+Here are the tests:
+
+```python
+def test_empty_knapsack():
+    assert knapsack([], [], 0) == 0
+
+def test_small_knapsack(self):
+    assert knapsack([10, 20, 30], [60, 100, 120], 50) == 220
+
+def test_medium_knapsack(self):
+    assert knapsack(
+        [99, 60, 70, 68, 50, 65, 48, 31, 52, 29],
+        [104, 105, 138, 135, 82, 73, 103, 52, 57, 136],
+        75
+    ) == 188
+```
+
+### !end-question
+### !placeholder
+
+```python
+def knapsack(weights, values, max_weight):
+    pass
+```
+### !end-placeholder
+### !tests
+```python
+import unittest
+from main import *
+
+class TestZeroOneKnapsack(unittest.TestCase):
+    def test_empty_knapsack(self):
+        self.assertEqual(knapsack([], [], 0), 0)
+
+    def test_small_knapsack(self):
+        self.assertEqual(knapsack([10, 20, 30], [60, 100, 120], 50), 220)
+
+    def test_medium_knapsack(self):
+        self.assertEqual(knapsack([99, 60, 70, 68, 50, 65, 48, 31, 52, 29], [104, 105, 138, 135, 82, 73, 103, 52, 57, 136], 75), 188)
+```
+### !end-tests
+### !hint
+
+- Is there a way we can think of this problem as resembling the the structure of the longest common subsequence problem?
+- If for the current item (weight and value) we take the weight and value, leaving the rest of the weights and values, what are the choices we can make about taking an item or leaving an item?
+- What limits are there on whether we are _allowed_ to take an item?
+
```

Comment:
I _think_ this is what I had in mind for the prompt I deleted, but this should be more clear.
```suggestion
- If we are _able_ to take an item, what aspects around taking or not taking the item do we need to consider?

```